refactor: changes for higher compatibility

fenpaws opened the pull request:
I wanted to generate a base API for Golang with the Swagger you provided, but the generator didn't understand the x-constants. We had to Fix a few things in the documentation to make it work,

- replaced x-constants with components
- added missing 200 response for POST request to upload whitelist

fenpaws committed:
refactor: changes for higher compatibility

- replaced x-components with components
- added missing 200 response for POST request to upload whitelist

fenpaws committed:
fix: Define integer type for relevant schema properties

- changed known ids/numbers to integers so no float or doubles are created where they are not needed for code generation

fenpaws committed:
fix: convert more numbers to integers & added missing optional properties

- found more numbers that cn be normal integers
- added more optional properties for the user

fenpaws committed:
fix: user endpoint accept strings instead of integers

The API endpoint allows for getting the user by ID and by their Username.

fenpaws committed:
fix: typo of current user property and change response at user endpoint to FullUser instead of User

DonovanDMC commented:
I'm going to leave this open for a bit longer since I assume you're implementing things based off of the spec and adding to this pr as you notice issues

DonovanDMC commented:
I've approved the workflows to run so linting issues show up, you can run it locally with `pnpm run lint:api`

fenpaws committed:
fix: liniting bugs

fenpaws committed:
fix: change all type number to integer if it is clearly an int

fenpaws committed:
fix: Update schema type of ip ban ID to integer in API definition

fenpaws commented:
I think we are done with the fixes.
We successfully generated a usable [Go Library](https://github.com/anthrove/openapi-e621-go) and your linter also run through without a problem.

Also, it should now be more in spec than before, this should increase the compatibly to doc generators. We didn't test it tho.

You should be able to merge it.

DonovanDMC commented on openapi.yaml:
```diff
@@ -561,7 +578,7 @@ components:
           type: boolean
         warning_type:
           type: ["string", "null"]
-          enum: *warning-types
+          "$ref": "#/components/schemas/WarningTypes"
```
Does the type in the component override the type set here?

alphyron replied:
Ah, yes. There is an error. The ref overwrites the  type which is set here. I will fix it.

alphyron committed:
Update openapi.yaml

fix: make warning type nullable at desired locations and change more numbers to integer

DonovanDMC commented on openapi.yaml:
```diff
@@ -2234,10 +2245,10 @@ components:
             - alias
         tagTypeA:
           type: ["number", "null"]
-          enum: *tag-categories
+          "$ref": "#/components/schemas/TagCategories"
         tagTypeB:
           type: ["number", "null"]
-          enum: *tag-categories
+          "$ref": "#/components/schemas/TagCategories"
```
these too

DonovanDMC commented on openapi.yaml:
```diff
@@ -2495,7 +2501,33 @@ components:
         avatar_id:
-          type: ["number", "null"]
+          type: ["integer", "null"]
+        artist_version_count:
+          type: integer
+        comment_count:
+          type: integer
+        favorites_count:
+          type: integer
+        flag_count:
+          type: number
+        forum_post_count:
+          type: integer
+        negative_feedback_count:
+          type: integer
+        neutral_feedback_count:
+          type: integer
+        pool_version_count:
+          type: integer
+        positive_feedback_count:
+          type: integer
+        profile_about:
+          type: string
+        profile_artinfo:
+          type: string
+        upload_limit:
+          type: integer
+        wiki_page_version_count:
+          type: integer
```
These are intentionally omitted, see `FullUser` and https://e621.net/users.json?search[id]=323290

DonovanDMC commented on openapi.yaml:
```diff
@@ -9922,6 +9931,10 @@ paths:
                   type: string
                 upload_whitelist[hidden]:
                   type: boolean
+      responses:
+        200:
+          description: Success
+
```
```suggestion
      responses:
        201:
          description: Success
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/UploadWhitelist"
        403:
          $ref: "#/components/responses/AccessDenied"
        422:
          $ref: "#/components/responses/ExpectedError"

```

DonovanDMC reviewed (changes requested):
Also some consistency issues, single quotes and quoted properties ("$ref")

alphyron committed:
fix: let TagCategories be null

alphyron committed:
fix: remove wrong fields

alphyron committed:
feat: define responses for POST upload_whitelists

alphyron committed:
fix: ref with quotes

alphyron commented:
> Also some consistency issues, single quotes and quoted properties ("$ref")

I need to say, I don't understand what do you mean by that... I search for lines with "$ref" where the value has single quotes but I can't find any... did you mean something else?

DonovanDMC commented:
> > Also some consistency issues, single quotes and quoted properties ("$ref")
> 
> I need to say, I don't understand what do you mean by that... I search for lines with "$ref" where the value has single quotes but I can't find any... did you mean something else?

They don't need quotes at all, no properties should the quoted unless they need to be

alphyron committed:
refactor: remove quotes from

alphyron commented:
Also, a while ago, we had the idea it would be interesting to have something like a test user at e621 with that we can execute some integration tests. Something like an additional validation task.
Of course there are some endpoints you cannot test like the admin or moderation stuff but the most used things like posts, user, faves and so on could be checked.
What are your thoughts about it?

DonovanDMC commented:
> Also, a while ago, we had the idea it would be interesting to have something like a test user at e621 with that we can execute some integration tests. Something like an additional validation task. Of course there are some endpoints you cannot test like the admin or moderation stuff but the most used things like posts, user, faves and so on could be checked. What are your thoughts about it?

It's trivial to run an instance of e621ng on your own

alphyron commented:
> > Also, a while ago, we had the idea it would be interesting to have something like a test user at e621 with that we can execute some integration tests. Something like an additional validation task. Of course there are some endpoints you cannot test like the admin or moderation stuff but the most used things like posts, user, faves and so on could be checked. What are your thoughts about it?
> 
> It's trivial to run an instance of e621ng on your own

Yeah is a point.


Except of that... Are there any other open points I should fix?

alphyron committed:
Implement contributor tags

fenpaws committed:
fix: Change `oneOf` to `anyOf` for response schema items

alphyron committed:
It only can one of the two schemas so change to oneOf

fenpaws committed:
feat: corrected types

- removed date-time format:  e621 sends more data then the format can hold
- fixed default_image_size typo
- api_regen_multiplier type to be a integer

fenpaws committed:
refactor: first carveout

fenpaws committed:
chore: removed acclimatised push

fenpaws committed:
feat: ignore some local testing files

fenpaws committed:
fix: remove the 'null' type when there is a ref

fenpaws committed:
feat: reimplemented artist urls

fenpaws committed:
feat: reimplemented avoid posting endpoints

fenpaws committed:
feat: reimplemented ban endpoints

fenpaws committed:
feat: reimplemented blip endpoints

fenpaws committed:
feat: reimplemented bulk update request endpoints

fenpaws committed:
feat: reimplemented comments endpoints

fenpaws committed:
feat: reimplemented last comment endpoints

fenpaws committed:
feat: reimplemented dmail endpoints

fenpaws committed:
feat: reimplemented dText endpoints

fenpaws committed:
feat: reimplemented email blacklists endpoints

fenpaws committed:
feat: reimplemented favorites endpoints

fenpaws committed:
feat: reimplemented forum endpoints

fenpaws committed:
feat: reimplemented forum endpoints

fenpaws committed:
feat: reimplemented forum topics endpoints

fenpaws committed:
fix: wrong ref

fenpaws committed:
feat: reimplemented help endpoints

fenpaws committed:
feat: reimplemented ip bans endpoints

fenpaws committed:
feat: reimplemented iqdb endpoints

fenpaws committed:
feat: reimplemented mascots endpoints

fenpaws committed:
feat: reimplemented mod action endpoints

fenpaws committed:
feat: reimplemented news update endpoints

fenpaws committed:
feat: reimplemented notes endpoints

fenpaws committed:
feat: reimplemented pools endpoints

fenpaws committed:
feat: reimplemented popular endpoints

fenpaws committed:
feat: reimplemented posts endpoints

fenpaws committed:
feat: reimplemented post sets endpoints

fenpaws committed:
feat: reimplemented post sets endpoints

fenpaws committed:
feat: reimplemented last post endpoints

fenpaws committed:
feat: reimplemented tag endpoints

fenpaws committed:
feat: reimplemented takedown endpoints

fenpaws committed:
feat: reimplemented tickets endpoints

fenpaws committed:
feat: reimplemented upload endpoints

fenpaws committed:
feat: reimplemented user endpoints

fenpaws committed:
feat: reimplemented wiki endpoints

fenpaws committed:
feat: reimplemented admin endpoints

Stevetec-exe committed:
feat: added bundling of openapi

fenpaws committed:
fix: linter errors

fenpaws committed:
feat: add a user response

fenpaws committed:
fix: remove idea folder

fenpaws committed:
feat: added TOS

fenpaws committed:
Merge pull request #1 from anthrove/restructure

OpenAPI Restructure

fenpaws commented:
Heya, we just did a huge refactor of the API, or rather how it's generated. All tests passed.

fenpaws committed:
fix: pnpm fix

DonovanDMC commented:
I'm going to assume this has been tested further beyond the linting passing, and that I don't need to spend several hours pouring over every detail?

fenpaws commented:
yeah i did :D It's basically just an extraction of the main OpenAPI you wrote with some fixes.

DonovanDMC commented:
It's failing to load, and the error in the console is completely incomprehensible

alphyron commented:
Intresting, because the finals openapi not changed alot, but yeah you are right. I can reproduce the error, when I run it local. I will check it.

DonovanDMC commented:
Seems likely to be a version mismatch in some dependency, my local repository which had not been updated with the dependency changes was able to run the webpack dev just fine, but when installing the newer dependencies it started showing the same issue seen in prod

alphyron commented:
so, webpack in the new version doesn't works? then whioch method you would prefer to fix  this problems? reverting dependencies to the old method or creating a new one for the ui?

fenpaws commented:
As stupid as it sounds, but it works on my machine. I cleaned all my NPM caches and did a clean `pnpm i` then running `pnpm run build:production && pnpm run start`. And i don't get any errors, it also opens perfectly fine.
<img width="1330" height="1490" alt="image" src="https://github.com/user-attachments/assets/fca7e65f-dab3-4b6d-9b9f-8337da19348b" />

DonovanDMC commented:
So it works fine with pnpm but it shits the bed without it
*sigh* I guess I'll bring back pnpm, seems like we're stuck with it because its version resolution is different in a way that works

DonovanDMC commented:
It indeed does work with pnpm, so looks like we're stuck with that
I really hate getting locked into a package manager like this but I'm not spending hours trying to figure out what's different about their dependency resolution that's making it work
